src/Plutip/Types.purs

```diff
@@ -197,27 +231,116 @@ instance DecodeAeson StopClusterResponse where
 -- | distribution for wallets.
 -- | Number of wallets in distribution specification matches the number of
 -- | wallets provided to the user.
-class UtxoDistribution distr wallets | distr -> wallets, wallets -> distr where
+class UtxoDistribution distr wallets | distr -> wallets where
   encodeDistribution :: distr -> Array (Array UtxoAmount)
-  decodeWallets :: Array PrivateKeyResponse -> Maybe wallets
+  decodeWallets :: distr -> Array PrivateKeyResponse -> Maybe wallets
+  keyWallets :: Proxy distr -> wallets -> Array KeyWallet
 
 instance UtxoDistribution Unit Unit where
   encodeDistribution _ = []
-  decodeWallets _ = Just unit
+  decodeWallets _ _ = pure unit
+  keyWallets _ _ = []
 
 instance UtxoDistribution InitialUTxO KeyWallet where
   encodeDistribution amounts = [ amounts ]
-  decodeWallets [ (PrivateKeyResponse key) ] =
-    pure $ privateKeysToKeyWallet (PrivatePaymentKey key) Nothing
-  decodeWallets _ = Nothing
+  decodeWallets _ [ PrivateKeyResponse key ] =
+    Just $ privateKeysToKeyWallet (PrivatePaymentKey key) Nothing
+  decodeWallets _ _ = Nothing
+  keyWallets _ wallet = [ wallet ]
+
+instance UtxoDistribution InitialUTxOWithStakeKey KeyWallet where
+  encodeDistribution (InitialUTxOWithStakeKey _ amounts) = [ amounts ]
+  decodeWallets (InitialUTxOWithStakeKey stake _) [ PrivateKeyResponse key ] =
+    Just $ privateKeysToKeyWallet (PrivatePaymentKey key) (Just stake)
+  decodeWallets _ _ = Nothing
+  keyWallets _ wallet = [ wallet ]
 
 instance
-  UtxoDistribution restSpec restWallets =>
-  UtxoDistribution (InitialUTxO /\ restSpec) (KeyWallet /\ restWallets) where
-  encodeDistribution (amounts /\ rest) =
-    encodeDistribution amounts <> encodeDistribution rest
-  decodeWallets = Array.uncons >>> case _ of
-    Nothing -> Nothing
-    Just { head: PrivateKeyResponse key, tail } ->
-      Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
-        decodeWallets tail
+  ( UtxoDistribution headSpec headWallets
+  , UtxoDistribution restSpec restWallets
+  ) =>
+  UtxoDistribution (headSpec /\ restSpec) (headWallets /\ restWallets) where
+  encodeDistribution (distr /\ rest) =
+    encodeDistribution distr <> encodeDistribution rest
+  decodeWallets (distr /\ rest) =
+    Array.uncons >=>
+      \{ head, tail } -> do
+        wallet <- decodeWallets distr [ head ]
+        Tuple wallet <$> decodeWallets rest tail
+  keyWallets _ (headWallets /\ restWallets) =
+    (keyWallets (Proxy :: Proxy headSpec) headWallets)
+      <> (keyWallets (Proxy :: Proxy restSpec) restWallets)
+
+type WalletInfo =
+  { utxos :: Utxo
+  , payPkh :: PaymentPubKeyHash
+  , stakePkh :: StakePubKeyHash
+  , wallet :: KeyWallet
+  }
+
+-- | For each wallet which includes a stake key, transfer the value of
+-- | the utxos at its enterprise address to its base address.
+transferFundsFromEnterpriseToBase
```

Comment:
Maybe this would fit better in a different file

Reply:
Let's move all `UtxoDistribution` machinery to `Plutip.UtxoDistribution`